Add validation checks for participations and missions requests

Floury-Fabio opened the pull request:
This branch add validation checks for admin missions requests and admin enrollment requests

For mission, it checks:
 - if related enrollments are not outside the new mission's period
 - if related enrollments on regulated mission match the mission's time slot

For enrollment, it checks:
- if the standard mission is full
- the slot availability for regulated mission
- the cash register proficiency for the regulated mission
- the positivity of the duration
- if the enrollment's date times are inside the period of the mission

We add also a drop down preset for creating enrollments, and fix display errors when rendering the response of a mission `#create` request.

Floury-Fabio committed:
add controll on datetimes of enrollments
It check if datetimes of enrollments are in datetime of his mission

Floury-Fabio committed:
add test for control enrollment params
- check positivity duration
- consistent of enrollment's datetimes

Floury-Fabio committed:
add test for update admin mission
this test check the control on the related enrollments before the
mission update

Floury-Fabio committed:
add controls for create enrollment admin request
it check:
  - if the standard mission is full
  - the slot availability for reulated mission
  - the cash register master for the regulated mission

Floury-Fabio committed:
add control for admin enrollment put on datetimes
It controls:
  - the positivity of the duration
  - if the enrollment's datetimes are inside the period of the mission

Floury-Fabio committed:
add control for admin enrollment put requests
it check:
  - if the time slots are available
  - if the cash register is insufficient when the is one slot left

Floury-Fabio committed:
fix:add update enrollment step in update transaction

Floury-Fabio committed:
add i18n keys for en

Floury-Fabio committed:
split update mission and recurrent update transactions
in order to prepare controlls in update mission transaction

Floury-Fabio committed:
add controls on put Mission
it check:
  - if related enrollments are not outside the new mission's period
  - if related enrollments regulated mission are match mission's the time slot

Floury-Fabio committed:
exclude blank from selection of enrollments params
this selection was permitted in admin form

Floury-Fabio committed:
fix: reactive errors displays for admin create request

Floury-Fabio commented on config/locales/models/enrollment.en.yml:
```diff
@@ -0,0 +1,12 @@
+    errors:
+      models:
+        enrollment:
+          inconsistent_datetimes: The selected datetimes are outside of the period of the mission
```
alphabetical order

Floury-Fabio committed:
fix spelling errors

Mate2xo commented on app/admin/enrollments.rb:
```diff
@@ -9,13 +9,71 @@
+            .new
+            .call(input)
+        end
+    end
```
Since you use `build_resource`, if the available `resource` helper already contains the associated `mission` and `member`, you could simplify this by doing something like
```ruby
def create
  build_resource
  transaction_result = Admin::Enrollments::CreateTransaction.new.call(resource)
  if transaction_result.success?
    flash[:notice] = translate 'enrollments.create.confirm_enroll'
    redirect_to admin_mission_path(params[:mission_id])
  else
    flash[:error] = transaction_result.failure
    render :new
  end
end
```

Mate2xo commented on app/admin/missions.rb:
```diff
@@ -83,9 +83,20 @@
+    def create # rubocop:disable Metrics/AbcSize
+      build_resource
+      if resource.save
+        flash[:notice] = translate 'missions.activerecord.notices.messages.record_created'
```
For all admin controller I18n lookups, do:
`translate '.message_that_you_what_to_lookup'`
This will use an I18n shortcut such like `admin.resource_name_pluralized.controller_action_name`

So in this case that would be for example:
`flash[:notice] = t '.success'`
and this would lookup `admin.missions.create.success`
and you could put the locale file in `config/locale/controllers/admin/missions.fr.yml`

Mate2xo commented on app/admin/missions.rb:
```diff
@@ -83,9 +83,20 @@
     def update
       if update_transaction.success?
-        flash[:notice] = translate 'activerecord.notices.messages.update_success'
+        flash[:notice] = translate 'missions.update.confirm_update'
```
Same here, try to use  the `t '.your_key'` shortcut where you can to help tidying locales with the default convention

Mate2xo commented on app/transactions/admin/missions/recurrent_update_transaction.rb:
```diff
@@ -0,0 +1,83 @@
+        result
+      end
+
+      def remove_datetime_attributes(input)
```
Leave a comment for why this step is necessary
e.g.:
`# For recurrent changes, we only update non time-related attributes to not mess with individual time slots validations. Time slots validation's responsibility belongs to each individual missions`

Mate2xo commented on app/transactions/admin/missions/recurrent_update_transaction.rb:
```diff
@@ -0,0 +1,83 @@
+      end
+
+      def build_mission_input(mission, params)
+        mission_params = mission.attributes
+        params.each do |key, value|
+          mission_params[key] = value
+        end
```
`params[:the_attribute_you_want] = mission.the_attribute_you_want`
Choose explicitly the attributes that you want to setup

Floury-Fabio committed:
refine test precisions
  - add helper which simulate datepicker params
  - define expected params for test all attributes in update request

Floury-Fabio committed:
add mission id in enrollment params for update request

Floury-Fabio committed:
transaction simplification
  - rm useless helpers from controllers
  - standardise input for update transaction
  - add helpers which manage datepicker transaction

Floury-Fabio committed:
transaction simplification
  - rm useless helpers from controllers
  - standardise input for update_transaction

Floury-Fabio committed:
add lazy lookup for i18n

Floury-Fabio committed:
add commentary in recurrent update transaction

Floury-Fabio committed:
fix spelling errors

Mate2xo commented on app/transactions/admin/enrollments/update_transaction.rb:
```diff
@@ -0,0 +1,180 @@
+        failure_message = I18n.t('activerecord.errors.models.enrollment.negative_duration')
+        return Failure(failure_message) unless input[:params][:start_time] < input[:params][:end_time]
+
+        Success(input)
+      end
+
+      def check_if_datetimes_of_enrollment_are_inside_the_mission_s_period(input)
+        failure_message = I18n.t('activerecord.errors.models.enrollment.inconsistent_datetimes')
+        return Failure(failure_message) unless are_inside_the_mission_s_period?(input)
+
+        Success(input)
+      end
+
+      def check_if_enrollment_are_matching_the_mission_s_timeslots(input)
+        return Success(input) unless input[:enrollment].mission.regulated?
+
+        failure_message = I18n.t('activerecord.errors.models.enrollment.time_slot_mismatch')
+        return Failure(failure_message) unless match_a_time_slot?(input)
+
+        Success(input)
+      end
+
+      def check_slot_availability_for_regulated_mission(input)
+        mission = input[:enrollment].mission
+        failure_message = I18n.t('activerecord.errors.models.enrollment.slot_unavailability')
+        return Success(input) if mission.genre != 'regulated'
+
+        return Failure(failure_message) unless are_all_time_slots_available?(input)
+
+        Success(input)
+      end
+
+      def check_cash_register_mastery(input)
+        return Success(input) unless input[:enrollment].mission.regulated?
+
+        failure_message = I18n.t('activerecord.errors.models.enrollment.insufficient_cash_register_mastery')
+        return Failure(failure_message) unless inspect_all_time_slots_for_register_cash_mastery(input)
+
+        Success(input)
+      end
```
Since this is duplicated from the related `create_transaction`, and this actually checks if an `Enrollment` is valid, you should extract these methods and add them as validation steps into the `Entrollment` model

Mate2xo commented on app/transactions/admin/missions/recurrent_update_transaction.rb:
```diff
@@ -0,0 +1,95 @@
+        { params: params, mission: mission }
+      end
+
+      def remove_datepicker_params(input, key)
```
`      def remove_activeadmin_datepicker_params(input, key)`

Mate2xo commented on app/transactions/admin/missions/update_transaction.rb:
```diff
@@ -6,72 +6,82 @@ module Missions
-
+      def check_if_enrollments_match_a_mission_s_time_slots_for_regulated_mission(input)
+        enrollments = input[:mission].enrollments
+        i18n_key = 'activerecord.errors.models.mission.mismatch_between_time_slots_and_related_enrollments'
+        failure_message = I18n.t(i18n_key)
```
```ruby
failure_message = I18n.t('mismatch_between_time_slots_and_related_enrollments',
                              scope: %i[activerecord errors models mission])
```

Mate2xo commented on app/transactions/admin/missions/update_transaction.rb:
```diff
@@ -6,72 +6,82 @@ module Missions
+          return Failure(failure_message) unless enrollment_inside_mission_s_period?(enrollment, input)
         end
-        result
+        Success(input)
       end
 
-      def remove_datetime_attributes_if_recurrent_option(input)
-        return Success(input) unless input[:params][:recurrent_change]
-
-        input[:params].delete(:start_date)
-        input[:params].delete(:due_date)
-
+      def check_if_enrollments_match_a_mission_s_time_slots_for_regulated_mission(input)
+        enrollments = input[:mission].enrollments
+        i18n_key = 'activerecord.errors.models.mission.mismatch_between_time_slots_and_related_enrollments'
+        failure_message = I18n.t(i18n_key)
+        enrollments.each do |enrollment|
+          return Failure(failure_message) unless match_a_time_slot?(input, enrollment)
+        end
         Success(input)
       end
```
THese methods are actually validating the `Mission` object ; so move them as validation steps in the `Mission` model.
By  doing so, we might be able to get rid of this transaction object.

Floury-Fabio commented on config/locales/controllers/enrollments.fr.yml:
```diff
@@ -6,6 +6,9 @@ fr:
       insufficient_proficiency: Votre maîtrise de la caisse est insuffisante pour prendre la dernière place du créneau de %{start_time} à %{end_time}
       max_member_count_reached: Le nombre maximum de participants est déjà atteint
       time_slots_requirement: Ils faut indiquer des créneaux fixes
```
'il faut'

Floury-Fabio committed:
fix spelling errors

Floury-Fabio committed:
rm useless rollback

Floury-Fabio committed:
move enrolment and mission validation in models

Mate2xo commented on app/models/enrollment.rb:
```diff
@@ -28,6 +28,90 @@ def contain_this_time_slot?(time_slot)
+    end
+
+    true
+  end
```
These methods should be `private`.
Then, to call them when you want to validate an `Enrollment` instance, use Rails' `validate` helper ; as in
```ruby
validate :check_cash_register_proficiency
``` 
Then, instead of asking your `enrollment` object in the `Admin::Enrollment::CreateTransaction` every `check_...` methods, you could just do `enrollment.valid?` to populate the `.errors`

Mate2xo commented on app/models/mission.rb:
```diff
@@ -89,6 +95,76 @@ def available_slots_count_for_a_time_slot(time_slot)
+      current_time_slot += 90.minutes
+    end
+    true
+  end
```
This might be better in the `Enrollment` model, as you will then ask each associated `enrollement` if they are still valid.
See also the Rails API documentation on the `has_many` options, as there are some validation-related options

Mate2xo commented on app/models/enrollment.rb:
```diff
@@ -28,6 +28,90 @@ def contain_this_time_slot?(time_slot)
+    ((end_time.to_i - start_time.to_i) % (60 * 90)).zero?
+  end
+
+  def check_if_datetimes_of_enrollment_are_inside_the_mission_s_period
+    unless mission.inside_period?(self)
+      failure_message = I18n.t('activerecord.errors.models.enrollment.inconsistent_datetimes')
+      errors.add :inconsistent_datetimes, failure_message
+      return false
+    end
```
Since an Enrollment belongs to a Mission, and you are asking here to the Mission if the current enrollment is valid; the responsiblity should be kept in the Enrollment.
You could refactor this as
```ruby
  def check_if_datetimes_of_enrollment_are_inside_the_mission_s_period
    unless inside_mission_period?
      failure_message = I18n.t('activerecord.errors.models.enrollment.inconsistent_datetimes')
      errors.add :inconsistent_datetimes, failure_message
      return false
    end
  end

  private

  def inside_mission_period?
    start_time >= mission.start_date &&
      start_time <= mission.due_date &&
      end_time >= mission.start_date &&
      end_time <= mission.due_date
  end

Mate2xo commented on app/transactions/admin/enrollments/create_transaction.rb:
```diff
@@ -0,0 +1,114 @@
+        return Failure(failure_message) unless slot_available_for_given_cash_register_proficiency?(input)
+
+        Success(input)
+      end
```
All these methods have already been moved to the `Enrollement`. You should delete them

Mate2xo commented on app/transactions/admin/enrollments/create_transaction.rb:
```diff
@@ -0,0 +1,114 @@
+        mastery_level_of_mission = Mission.cash_register_proficiency_requirements[proficiency_requirement]
+
+        mastery_level_of_member >= mastery_level_of_mission
+      end
```
It seems that this method isn't called anywhere?

Mate2xo commented on app/models/mission.rb:
```diff
@@ -89,6 +95,76 @@ def available_slots_count_for_a_time_slot(time_slot)
+      end
+    end
+    true
+  end
```
Move them to `private`, and call them through the `validate :your_method` Rails helper.
You could then call `.valid` on your `Mission` instance in `Admin::Missions::UpdateTransaction`

Mate2xo commented on app/transactions/admin/enrollments/update_transaction.rb:
```diff
@@ -0,0 +1,67 @@
+
+        params.keys.each do |current_key| # rubocop:disable Style/HashEachMethods
+          enrollment[current_key] = params[current_key] if enrollment.attributes.keys.include?(current_key)
+        end
```
Are you trying to do the same thing as `.assign_attributes` here ?

Mate2xo commented on app/transactions/admin/enrollments/update_transaction.rb:
```diff
@@ -0,0 +1,67 @@
+        enrollment.check_if_datetimes_of_enrollment_are_inside_the_mission_s_period
+        enrollment.check_if_enrollment_is_matching_the_mission_s_timeslots
+        enrollment.check_slots_availability_for_regulated_mission
+        enrollment.check_cash_register_proficiency
```
Use `.valid?` instead

Mate2xo commented on app/transactions/admin/missions/update_transaction.rb:
```diff
@@ -6,72 +6,59 @@ module Missions
+
+        params.keys.each do |current_key| # rubocop:disable Style/HashEachMethods
+          mission[current_key] = params[current_key] if mission.attributes.keys.include?(current_key)
         end
```
`mission.assign_attributes(params)`

Mate2xo commented on app/transactions/admin/missions/update_transaction.rb:
```diff
@@ -6,72 +6,59 @@ module Missions
+        mission = input[:mission]
+        mission.check_if_enrollments_are_inside_new_mission_s_period
+        mission.check_if_enrollments_match_a_mission_s_time_slots_for_regulated_mission
+        return Failure(mission.errors.values.flatten[0]) if mission.errors.present?
```
use `.valid?`

Mate2xo commented on config/locales/models/mission.en.yml:
```diff
@@ -16,6 +16,8 @@ en:
       models:
         mission:
+          inconsistent_datetimes_for_related_enrollments: At last one enrollment is outside the mission's period
+          mismatch_between_time_slots_and_related_enrollments: Mismatch_between_time_slots_and_related_enrollments
```
The `recurrent_update_cancel` key is missing (compared to the .fr locale)

Mate2xo commented on spec/requests/admin/enrollments_spec.rb:
```diff
@@ -0,0 +1,390 @@
+    end
+  end
+
+  # this helper imitate the struture of date_params created by active admin datepicker
+  def convert_datetime_in_params(datetime, key)
+    {
+      "#{key}(1i)": datetime.year,
+      "#{key}(2i)": datetime.month,
+      "#{key}(3i)": datetime.day,
+      "#{key}(4i)": datetime.hour,
+      "#{key}(5i)": datetime.min
+    }
+  end
```
I think I saw this same helper at least 3 times. You could maybe extract it in its own service object

Mate2xo commented on spec/requests/admin/enrollments_spec.rb:
```diff
@@ -0,0 +1,390 @@
+
+require 'rails_helper'
+
+RSpec.describe 'A Enrollment admin request', type: :request do
```
Since all the nested `describe` are HTTP method, I think that having the main parent `describe` as 
```ruby
RSpec.describe 'admin/enrollments/', type: :request do
```
would be more explicit. What do you think?

Mate2xo commented on spec/requests/admin/enrollments_spec.rb:
```diff
@@ -0,0 +1,390 @@
+      expect(CGI.unescapeHTML(response.body)).to include(I18n.t('enrollments.create.confirm_enroll'))
+    end
+
+    context 'when the related mission is :regulated and the enrollment match a time slot' do
```
context 'when the related mission is :regulated and the enrollment match`es` a time slot' do

Mate2xo commented on spec/requests/admin/enrollments_spec.rb:
```diff
@@ -0,0 +1,390 @@
+    end
+
+    context 'when the related mission is regulated, there is one slot left,
+             and the member has an insufficient cash register mastery' do
```
mastery -> proficiency

Mate2xo commented on spec/requests/admin/missions_spec.rb:
```diff
@@ -29,32 +28,241 @@
+        other_missions = create_future_matching_missions(mission)
 
         put_mission
 
```
no expectation?

Floury-Fabio committed:
move cash register proficiency check in validator

for this we have also did:
  - rewrite several tests broken by new validator
  - rm cash register proficiency check from enrollment create
  transaction
  - reorganize some i18n keys

Floury-Fabio committed:
move the inclusion enrollment'datetimes check in validator

with this we have also did:
  - adapt several test because they was broken by validators
  - simplify enrollment factory
  - rm this check from enrllment transaction
  - set callback #set_defaut to before_validation in order to
   adapt at new validators

Floury-Fabio committed:
move uniquness enrollment check in validator

we had also adapt several tests breken by the new validator

Floury-Fabio committed:
move availability_place check in validator

this validator check if the mission have a place before the enrollment

Floury-Fabio committed:
move duration check for enrollment in validator

Floury-Fabio committed:
move validators in modules

Floury-Fabio committed:
rm useless validation method from transaction
this check is in validator

Floury-Fabio committed:
move slots availability check in validators

Floury-Fabio committed:
move matching mission time slots check in validators

we have also move the 90 minutes duration check in duration validator

Floury-Fabio committed:
add test for 90 minutes duration validation

Floury-Fabio committed:
rm useless create enrollment transaction

Floury-Fabio committed:
refact: rm useless methods
this method was in admin enrollment update_transaction
it was become useless because his function is ensured by validator

Floury-Fabio committed:
refact enrollment model
rm useless methods

Floury-Fabio committed:
add comment on admin enrollment controllers

Floury-Fabio committed:
check related enrollments with validated_associates
it do also :
  -rm check from model and transaction
  - adapt request test to the changement
  - add i18n key

Note: the current local langage is not loaded in validated_associates
so the error message is displayed in english and the test doesnt check
the error message content

Floury-Fabio committed:
fix flaky test

Floury-Fabio committed:
fix i18n call

Floury-Fabio committed:
rm useless transaction and useless transaction

Floury-Fabio committed:
rm useless transaction

Floury-Fabio commented on app/admin/missions.rb:
```diff
@@ -83,29 +83,33 @@
+        flash[:notice] = translate '.mission_created'
+        redirect_to admin_mission_path(resource.id)
+      else
+        flash[:error] = resource.errors.full_messages.join
```
resource.errors.full_messages.join(', ')

Floury-Fabio committed:
fix spelling errors

Floury-Fabio committed:
fix spelling error again

Commit:
Merge branch 'development' into add_controls_for_participations_and_missions_requests

Commit:
Update Solargraph and Rubocop

Commit:
Improve comments

Commit:
Fix locale language mixup

Commit:
Change wording
